### PR TITLE
Add CircuitDiagramInfoArgs.format_real

### PR DESCRIPTION
**Pull request by Strilanc**
- Also fix the qasm UGate not having a decomposition or a correct repr

### CODE REVIEW

File: cirq/protocols/circuit_diagram_info_protocol.py [hunk 1]
```diff
@@ -133,6 +134,15 @@ def __repr__(self):
                                          self.use_unicode_characters,
                                          self.precision, self.qubit_map))
 
+    def format_real(self, val: Union[sympy.Basic, int, float]):
+        if isinstance(val, sympy.Basic):
+            return str(val)
```
Comment (viathor) [inline cirq/protocols/circuit_diagram_info_protocol.py]: Optional: check that `val` is real.

File: cirq/protocols/circuit_diagram_info_protocol.py [hunk 2]
```diff
@@ -133,6 +134,15 @@ def __repr__(self):
                                          self.use_unicode_characters,
                                          self.precision, self.qubit_map))
 
+    def format_real(self, val: Union[sympy.Basic, int, float]):
```
Comment (viathor) [inline cirq/protocols/circuit_diagram_info_protocol.py]: `-> str`

File: cirq/protocols/circuit_diagram_info_protocol.py [hunk 3]
```diff
@@ -133,6 +134,15 @@ def __repr__(self):
...
+        if isinstance(val, sympy.Basic):
+            return str(val)
+        if val == int(val):
+            return str(int(val))
+        if self.precision is None:
+            return str(val)
+        return '{{:.{}}}'.format(self.precision).format(float(val))
```
Comment (viathor) [inline cirq/protocols/circuit_diagram_info_protocol.py]: No need to resort to calling `format`:

```python
return f'{float(val):.{self.precision}f}'
```

should work.

File: cirq/circuits/qasm_output.py [hunk 4]
```diff
@@ -54,24 +54,26 @@ def from_matrix(mat: np.array) -> 'QasmUGate':
...
-        return 'cirq.QasmUGate({}, {}, {})'.format(self.theta, self.phi,
-                                                   self.lmda)
-
-    def _unitary_(self) -> np.ndarray:
-        # Source: https://arxiv.org/abs/1707.03429 (equation 2)
-        operations = [
-            ops.Rz(self.phi * np.pi),
-            ops.Ry(self.theta * np.pi),
-            ops.Rz(self.lmda * np.pi),
+        return (f'cirq.circuits.qasm_output.QasmUGate('
```
Comment (viathor) [inline cirq/circuits/qasm_output.py]: Maybe import `QasmUGate`  into top level namespace instead? The name is unlikely to collide and the repr would look a lot better.

File: cirq/protocols/circuit_diagram_info_protocol.py (hunk 2 above)
Comment (Strilanc) [reply cirq/protocols/circuit_diagram_info_protocol.py]: Done.

File: cirq/protocols/circuit_diagram_info_protocol.py (hunk 1 above)
Comment (Strilanc) [reply cirq/protocols/circuit_diagram_info_protocol.py]: Acknowledged.

File: cirq/protocols/circuit_diagram_info_protocol.py (hunk 3 above)
Comment (Strilanc) [reply cirq/protocols/circuit_diagram_info_protocol.py]: Much nicer. Done.

File: cirq/circuits/qasm_output.py (hunk 4 above)
Comment (Strilanc) [reply cirq/circuits/qasm_output.py]: Will leave it here for now. Can be done in separate change if needed.

**Pull request merged**